datumaro/plugins/voc_format/extractor.py

```diff
@@ -163,116 +197,141 @@ def __iter__(self):
+                    width = self._parse_field(root_elem, "size/width", int, required=False)
                     if height and width:
                         size = (height, width)
+
                     filename_elem = root_elem.find("filename")
                     if filename_elem is not None:
                         image = osp.join(image_dir, filename_elem.text)
-                    anns = self._parse_annotations(root_elem)
-                else:
+
+                    anns = self._parse_annotations(root_elem, item_id=(item_id, self._subset))
+
+                if image is None:
                     image = images.pop(item_id, None)
 
                 if image or size:
                     image = Image(path=image, size=size)
 
                 yield DatasetItem(id=item_id, subset=self._subset, media=image, annotations=anns)
+            except ElementTree.ParseError as e:
+                readable_wrapper = InvalidAnnotationError("Failed to parse XML file")
+                readable_wrapper.__cause__ = e
+                self._ctx.error_policy.report_item_error(
+                    readable_wrapper, item_id=(item_id, self._subset)
+                )
             except Exception as e:
-                self._report_item_error(e, item_id=(item_id, self._subset))
-
-    def _parse_annotations(self, root_elem):
-        item_annotations = []
-
-        for obj_id, object_elem in enumerate(root_elem.findall("object")):
-            obj_id += 1
-            attributes = {}
-            group = obj_id
-
-            obj_label_id = None
-            label_elem = object_elem.find("name")
-            if label_elem is not None:
-                obj_label_id = self._get_label_id(label_elem.text)
+                self._ctx.error_policy.report_item_error(e, item_id=(item_id, self._subset))
 
-            obj_bbox = self._parse_bbox(object_elem)
-
-            if obj_label_id is None or obj_bbox is None:
-                continue
-
-            difficult_elem = object_elem.find("difficult")
-            attributes["difficult"] = difficult_elem is not None and difficult_elem.text == "1"
-
-            truncated_elem = object_elem.find("truncated")
-            attributes["truncated"] = truncated_elem is not None and truncated_elem.text == "1"
-
-            occluded_elem = object_elem.find("occluded")
-            attributes["occluded"] = occluded_elem is not None and occluded_elem.text == "1"
+    @staticmethod
+    def _parse_field(root, xpath: str, cls: Type[T] = str, required: bool = True) -> Optional[T]:
+        elem = root.find(xpath)
+        if elem is None:
+            if required:
+                raise MissingFieldError(xpath)
+            else:
+                return None
 
-            pose_elem = object_elem.find("pose")
-            if pose_elem is not None:
-                attributes["pose"] = pose_elem.text
+        if not issubclass(cls, str):
```

Comment:
This condition is incorrect. If `cls` is a subclass of `str` (but not `str` itself), then you _do_ need to cast `elem.text` to `cls`, otherwise it'll be of the wrong type.

So the condition needs to be `if cls is not str`. Actually, for readability it might be better to rewrite it as:

```python
if cls is str:
    return elem.text

try:
    return cls(elem.text)
except Exception as e:
    raise InvalidFieldError(xpath) from e
```

Reply:
Ok, makes sense